Remove incremental

SmileyChris opened the pull request:
# Description
Remove Incremental as a requirement (but still support its use for project versions/names)

Fixes #308 and #398

# Checklist
<!-- add a "X" inside the brackets to confirm -->
* [x] Make sure changes are covered by existing or new tests.
* [x] For at least one Python version, make sure local test run is green.
* [x] Create a file in `src/towncrier/newsfragments/`. Describe your
  change and include important information. Your change will be included in the public release notes.
* [x] Make sure all GitHub Actions checks are green (they are automatically checking all of the above).
* [x] Ensure `docs/tutorial.rst` is still up-to-date.
* [x] If you add new **CLI arguments** (or change the meaning of existing ones), make sure `docs/cli.rst` reflects those changes.
* [x] If you add new **configuration options** (or change the meaning of existing ones), make sure `docs/configuration.rst` reflects those changes.

chrysle commented on src/towncrier/test/test_project.py:
```diff
@@ -48,25 +41,61 @@ def test_tuple(self):
+__version__ = WeirdVersion()
+"""
+            )
+        with self.assertRaises(Exception) as e:
```
Which exception does this raise?

SmileyChris replied:
Literally [`Exception`](https://github.com/twisted/towncrier/pull/491/files#diff-ea234c597bb51dfad3d154ebdeb24af6c75e95ac9a922267f99e7704f1fbde42R72) -- the definition of that exception is existing code, but we can of course tighten this up if it makes sense now.

adiroiban replied:
I think that it's ok to keep Exception for now.

Later we can look at raising ClickException or something like this for better CLI integration... but in a separate PR.

adiroiban commented:
Thanks, Chris for working on this.

I am working to make a new release of towncrier, and the plan is to drop 3.7 support.

So this PR should be even simpler.

I am happy to drop the `incremental` dependency from towncrier.
I think that managing it via a simple string is good enough.
The version should only be modified by the release manager and the release manager should be familiar with the versioning scheme.

I think that using `Version('towncrier', 19, 9, 0, release_candidate=1)` instead of just `19.9.0.rc1` is overengineering

SmileyChris commented:
@adiroiban cool! I'd say merge this first then work on dropping the 3.7 support, otherwise I can update this after you've changed the automated tests so this doesn't fail when I remove the import try/excepts.

(also a review of #482 would be nice, it would be a good addition before the new release ;) )

adiroiban commented on src/towncrier/_shell.py:
```diff
@@ -13,14 +13,25 @@
 from .create import _main as _create_cmd
 
 
+try:
```
Maybe we can still keep the `_version.py` file but this time just just a `_version` string variable.

I this way, we don't have to repeat the `importlib` code each time we want to fetch the version

hynek replied:
We can do this lazily using `towncrier.__version__` if the block that I've commented on earlier is ported and kill two birds with one stone.

sigma67 replied:
`_version.py` restored with contents

```
__version__ = "23.11.1.dev0"
```

adiroiban commented on src/towncrier/_shell.py:
```diff
@@ -13,14 +13,25 @@
+    from importlib_metadata import PackageNotFoundError, version  # type: ignore
+
+
+try:
```
this code can be included in the updated `_version.py` code

adiroiban commented on src/towncrier/test/test_packaging.py:
```diff
@@ -1,22 +1,16 @@
         """
-        Import __version__ from towncrier returns an Incremental version object
-        and raises a warning.
+        towncrier.__version__ was deprecated, now no longer exists.
```
not sure if `towncrier.__version__` was considered public or private API.

For private API, raising AttributeError is ok.
If it's public API, we need to raise a separate error with an info message.

adiroiban commented on src/towncrier/test/test_project.py:
```diff
@@ -46,25 +39,61 @@ def test_tuple(self):
     def test_incremental(self):
         """
-        An incremental Version __version__  is picked up.
+        An incremental-like Version __version__  is picked up.
```
what does `incremental-like` version means ?

I guess that this means a `__version__` object/instance that has the same public API/interface as incremental.
That is, it doesn't have to be `incremental` it only needs to behave like it.

SmileyChris replied:
Correct. The code doesn't check the instance of the __version__, just that it quacks like an incremental version.

adiroiban commented on src/towncrier/test/test_project.py:
```diff
@@ -46,25 +39,61 @@ def test_tuple(self):
+        with open(os.path.join(temp, "mytestprojinc", "__init__.py"), "w") as f:
+            f.write(
+                """
+class Version:
```
maybe a docstring can be added to infom that this tries to "emulate" an incremntal class.

adiroiban reviewed (changes requested):
The changes look good.

Only minor comments.

I think  that it would be nice to be able to still do `from towncrier._version import version` or something similar, without having to worry about importlib each time.

Now, I don't know if it should raise an exception or return a dummy string, when the version was not found... I would go with a "dummy/placeholder" string.

adiroiban commented:
@SmileyChris #482 it's on my todo list... but we can trigger a release anytime.

hynek commented:
JFTR: if you'd like to entertain the idea of using `hatch-vcs` to deduplicate version information, I'm happy to help. I used to be skeptical to the idea, but now I've moved all my projects to it. If configured correctly, it also allows for continuous uploads to Test PyPI: https://test.pypi.org/project/structlog/#history

I guess the "you" here is mostly towards @adiroiban since he's doing all the release work and @glyph because he's The Twisted King.

adiroiban commented:
To me `hatch-vcs` is over-engineering.
I am not sure what problem it tries to solve.

Is it that hard to keep the version as a text inside a .toml file? :)

hynek commented:
Well, it's not. :) But it deduplicates the version information to one place (we do git tags too) and enables the continuous uploads I've mentioned. But obviously, if you don't want it, it's not happening. Just wanted to offer. :)

hynek commented:
this needs a rebase / conflict resolution

SmileyChris commented:
Removed the python 3.7 related fallbacks, updated the version in the toml, and rebased against trunk.
(and pushing again with no change because the `checknewsfragment` test failed for no good reason 😖 )

SmileyChris committed:
Add news fragment

SmileyChris committed:
Remove incremental

SmileyChris committed:
Update tutorial for accuracy around detecting versions

hynek committed:
Merge branch 'trunk' into remove-incremental

hynek commented on src/towncrier/__init__.py:
```diff
@@ -4,29 +4,3 @@
-__all__ = ["__version__"]
-
-
-def __getattr__(name: str) -> Version:
-    if name != "__version__":
-        raise AttributeError(f"module {__name__} has no attribute {name}")
-
-    import warnings
-
-    from ._version import __version__
-
-    warnings.warn(
-        "Accessing towncrier.__version__ is deprecated and will be "
-        "removed in a future release. Use importlib.metadata directly "
-        "to query for towncrier's packaging metadata.",
-        DeprecationWarning,
-        stacklevel=2,
-    )
-
-    return __version__
```
I'm not sure this should be removed. It seems like there's still a bit of software relying on `pkg.__version__` such that I've starte de-deprecating it in my own packages. See also https://github.com/python-attrs/attrs/issues/1136

(to be clear: it has to be ported, not just left around ;))

sigma67 replied:
re-added #627

hynek commented on src/towncrier/_project.py:
```diff
@@ -38,47 +38,45 @@ def _get_package(package_dir: str, package: str) -> ModuleType:
 
 def get_version(package_dir: str, package: str) -> str:
-    module = _get_package(package_dir, package)
+    version: Any
```
Why Any? Any opts `version` completely out of type checking, is there a rescue for that?

sigma67 replied:
changed to str #627

hynek commented on src/towncrier/_project.py:
```diff
@@ -38,47 +38,45 @@ def _get_package(package_dir: str, package: str) -> ModuleType:
-
     version = getattr(module, "__version__", None)
+    # Incremental has support for package names, try duck-typing it.
+    try:
+        return str(version.package)  # type: ignore
+    except AttributeError:
+        pass
```
```suggestion
    with contextlib.suppress(AttributeError):
        return str(version.package)  # type: ignore
```

is more idiomatic, I think

sigma67 replied:
done  #627

hynek commented on src/towncrier/test/test_packaging.py:
```diff
@@ -1,22 +1,16 @@
 
 class TestPackaging(TestCase):
-    def test_version_warning(self):
+    def no_version_attr(self):
```
I don't think this runs if it doesn't have a `test_` prefix.

sigma67 replied:
changed test to check for deprecationwarning instead #627

hynek reviewed (changes requested):
I've added a few notes, but ultimately I'm unconditionally yielding the decisions to @adiroiban since this has been his dream for years. ;)

adiroiban commented on src/towncrier/_project.py:
```diff
@@ -38,47 +38,45 @@ def _get_package(package_dir: str, package: str) -> ModuleType:
-
     if not version:
-        raise Exception("No __version__, I don't know how else to look")
+        try:
```
it looks like this is not only about removing incremental, but also about using `importlib_metadata` to get the version of  the target project

I think is best to do this as part of a separate PR ..l like https://github.com/twisted/towncrier/pull/502

sigma67 replied:
Stale comment as linked PR has been merged. Updated and integrated latest trunk in  #627

sigma67 commented:
Hi @adiroiban @SmileyChris is this PR stale and ready to be picked up by someone else? Or are you planning to continue work?

twm commented on src/towncrier/_project.py:
```diff
@@ -38,47 +38,45 @@ def _get_package(package_dir: str, package: str) -> ModuleType:
+            version = str(version.base()).strip()
+            # Incremental uses `X.Y.rcN`.
+            # Standardize on importlib (and PEP440) use of `X.YrcN`:
+            return version.replace(".rc", "rc")  # type: ignore
```
FYI this has been fixed on the Incremental side.

sigma67 replied:
Not sure what you are referring to. Should anything in this PR be changed as a result?

adiroiban commented:
Hi @sigma67 

This PR is stale.

If you have time, feel free to continue working on it.

Thanks

sigma67 commented:
Hi @adiroiban, thanks, I've picked it up in #627

adiroiban commented:
I am closing this as the current effort is with https://github.com/twisted/towncrier/pull/627
